### ISSUE TITLE
Running Youtube-DL on Linux (Ubuntu)

### ISSUE THREAD

**Issue by spaceman5** [ytdl-org/youtube-dl#31810]
Hi all

Since youtube-dl.exe is not updated,
I tried installing youtube-dl on a Linux machine that I have (Ubuntu), using:
**apt install youtube-dl**

The installation was successful,
however when running it, the **uploader_id** problem remains there too.
I thought it would be updated there.

So my question is:  what should I do to get the most updated build there, that will work?

Maybe install Python, and then inside it install youtube-dl using pip?
Please suggest the simplest solution..

Thank you

**Comment by memberwebm** [comment]
I have the same problem, can you tell me how to update youtube dl?

**Comment by dirkf** [comment]
#30839.

**Comment by nicolaasjan** [comment]
> Maybe install Python, and then inside it install youtube-dl using pip?

Ubuntu already has Python. 🙂

`pip install git+https://github.com/ytdl-org/youtube-dl.git@master`

**Comment by dirkf** [comment]
Follow the link from the **pinned issue** (that you should have read already) which should have caused you to read the **actual issue** that covers this problem instead of opening a new issue **for no good reason at all**. In the master issue, you are advised how to update the program. In particular, you could add a PPA to your Ubuntu setup that has a new version of yt-dl. But since repo versions of yt-dl generally don't get updated often enough, even when yt-dl releases are made, the solutions using _pip_ may be better.

**Comment by spaceman5** [comment]
I did it, and it worked.
Thank you very much.

BTW,
what's the difference between:
pip install git+https://github.com/ytdl-org/youtube-dl.git@master#egg=youtube_dl
and
pip install git+https://github.com/ytdl-org/youtube-dl.git@master
?

What does the "#egg=youtube_dl" parameter do?
(I ran mine with the egg parameter)

**Comment by dirkf** [comment]
Please find out and tell us (in the master thread). A guess is that a package installable with _pip_ can have optional features and this syntax specifies the `youtube_dl` feature. Possibly there is no such feature and the parameter is silently ignored, or it might install a `youtube-dl` command, so that omitting it might not be important when updating a previous installation, or ...

**Comment by spaceman5** [comment]
Thank you dirkf.

BTW,
If I want to install the youtube-dl, but without using pip,
(and assuming I have python3 installed and working well on a Linux machine),
Can anyone please tell me how should I do it?

Is it as simple as
1) Clone the Git repo,
2) run python3 youtube-dl.py?

**Comment by nicolaasjan** [comment]
You will need the build tools: `zip`, `make` (GNU), `pandoc` and `pytest`.
(not sure about the last one)

Then:
`git clone https://github.com/ytdl-org/youtube-dl.git`
`cd youtube-dl`
`make`

Then  the executable will be right there in your `youtube-dl` directory. 🙂
Copy it to `/usr/local/bin`.

**Comment by spaceman5** [comment]
Thank you very much nicolaasjan.

And If I do it this way, using git clone,
and after some months the repository was updated,
then on my side, in order to update, what I should do is git pull?
(and then make again)
Anything else in addition to git pull, and make?

**Comment by nicolaasjan** [comment]
> Thank you very much nicolaasjan.
> 
> And If I do it this way, using `git clone`, and after some months the repository was updated, then on my side, in order to update, what I should do is `git pull`? (and then `make` again)
> 
> Anything else in addition to `git pull`, and `make`?

Never tried it with `git pull`...
I just delete the directory and start fresh. 😀️

**Comment by spaceman5** [comment]
Really?
Is that what most people do when they want to update their local copy of a repository to reflect all updates that exist on github?

**Comment by nicolaasjan** [comment]
No idea what most people do.
It's just a weird habit of mine. 😀️
Not really a problem, since it's not a very large repo.

But I'll try next time.

**Comment by spaceman5** [comment]
BTW,
I am experimenting with installing it the way you described
(`git clone`, and then `make`, instead of using `pip`),
and I made sure to install `make` and `zip` prior to starting.
when I cd youtube-dl,
and run `make`, this is what I get:

**Comment by nicolaasjan** [comment]
`sudo apt install python-is-python3` ?

**Comment by spaceman5** [comment]
You're right..
Thank you.

OK the make process completed successfully.

Here is the full list of what I used:

`apt install make`
`apt install zip`
`apt install pandoc`
`apt install python-is-python3`

Now qwhen I'm inside the youtube-dl folder,
If I run `youtube-dl`,
this is what I get:

> Command 'youtube-dl' not found, but can be installed with:
> apt install youtube-dl

(the file has the x attribute..)


Here is a screenshot:
https://i.postimg.cc/pdLf169w/yt.png

**Comment by spaceman5** [comment]
OK
Running `./youtube-dl` solved it..

I will add it to the PATH in ordeer not to have to prefix it with`./` all the time.

Thank you very much Nicholas

**Comment by nicolaasjan** [comment]
> Now qwhen I'm inside the youtube-dl folder,
> If I run youtube-dl,
> > Command 'youtube-dl' not found, but can be installed with:
>> apt install youtube-dl

In that folder, you'll have to run `./youtube-dl <URL>`.

Or copy it to `/usr/local/bin` and then you can run `youtube-dl <URl>` from anywhere.

**Comment by spaceman5** [comment]
Right..
thank you so much man,
I must say that the `pip` option is definitely much more fool proof,
yet here I learned abit more..
Thank you so much

**Comment by spaceman5** [comment]
> Or copy it to `/usr/local/bin` and then you can run `youtube-dl <URl>` from anywhere.

I think it might be better to create a Link in `/usr/local/bin`,
rather than copying the executable there,
that way, If in the future I delete the copy of the repo, and redownload it (and re-make it),
the link will point to an updated executable, instead of being the old executable..

**Comment by nicolaasjan** [comment]
Good idea.

**Comment by spaceman5** [comment]
Glad I managed to make a small contribution here too :))

BTW,
I now did a test run on a Youtube Video,
and it seems that despite the fact that youtube-dl is running,
there are now problems during the run.

> Traceback (most recent call last):
>   File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
>     exec(code, run_globals)
>   File "/root/youtube-dl/./youtube-dl/__main__.py", line 19, in <module>
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/__init__.py", line 475, in main
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/__init__.py", line 465, in _real_main
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/YoutubeDL.py", line 2083, in download
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/YoutubeDL.py", line 812, in extract_info
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/YoutubeDL.py", line 819, in wrapper
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/YoutubeDL.py", line 840, in __extract_info
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/extractor/common.py", line 535, in extract
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/extractor/youtube.py", line 2167, in _real_extract
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/extractor/youtube.py", line 447, in _extract_author_var
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/utils.py", line 1972, in get_element_by_attribute
>   File "/root/youtube-dl/./youtube-dl/youtube_dl/utils.py", line 1989, in get_elements_by_attribute
>   File "/usr/lib/python3.10/re.py", line 247, in finditer
>     return _compile(pattern, flags).finditer(string)
> TypeError: expected string or bytes-like object



So I think I will abandon the git clone + make method,
and revert back to pip.
The pip option is definitely easier.

You really need to be able to solve problems in Linux, in order to use the `make` option,
and I am not there yet..

I learned things today,
but I also learned that using `pip` should be preferred over `make`, when possible..

**Comment by spaceman5** [comment]
I created a completely new VPS,
and now did the simpler `pip` installation method that worked well previously.

Quite ironically,
but I now get the same error message..

> Downloading webpage
> WARNING: Unable to download webpage: HTTP Error 429: Too Many Requests

So is there maybe today a global problem for everyone?

Retrying to download,
leads to the same **HTTP Error 429: Too Many Requests**

**Comment by spaceman5** [comment]
It might be that the IP of that VPS is blocked by youtube..
Because on another computer that I have at home, youtube-dl works.

**Comment by nicolaasjan** [comment]
> I now did a test run on a Youtube Video,
> and it seems that despite the fact that youtube-dl is running,
> there are now problems during the run.

What video?

I would still like to know what caused this Python error with youtube-dl made by `make`...

```
File "/usr/lib/python3.10/re.py", line 247, in finditer
return _compile(pattern, flags).finditer(string)
TypeError: expected string or bytes-like object
```
(pinging @dirkf)

**Comment by spaceman5** [comment]
Any video..
They all got **HTTP Error 429: Too Many Requests** on that VPS.
It's probably an IP (or IP Range) blocked by Youtube.

When I try to download on a local Linux machine, it works well without Error 429..

(so maybe my git clone + make installation was perfectly fine)

**Comment by dirkf** [comment]
429 -> #30839.

In https://github.com/ytdl-org/youtube-dl/issues/31810#issuecomment-1471657189, I guess that there must have been an earlier `WARNING` caused by this line extractor/youtube.py:1839 failing:
```
        webpage = self._download_webpage(
            webpage_url + '&bpctr=9999999999&has_verified=1', video_id, fatal=False)
```
After a (eg) 429 error in line 1839, trying to `_extract_author_var()` from a `None` `webpage` causes the crash.

At line 447, this would make it safe:
```diff
-                get_element_by_attribute('itemprop', 'author', webpage) or '',
+                get_element_by_attribute('itemprop', 'author', webpage or '') or '',
```

**Comment by spaceman5** [comment]
In case this info is useful for you,
I tested in the past several hours,
on several newly created VPSes,
with different IP Addresses, and in different countries,
and also installed both using pip and using make,
and all, have the Error 429 problem.

Is it possible that it's not IP Blocking by google, and something else instead?
